shakenfist/baseobject.py

```diff
@@ -0,0 +1,105 @@
+
+from functools import partial
+import time
+
+from shakenfist import db
+from shakenfist import etcd
+from shakenfist import exceptions
+from shakenfist import logutil
+
+
+LOG, _ = logutil.setup(__name__)
+
+
+class DatabaseBackedObject(object):
+    object_type = 'unknown'
+
+    def __init__(self, object_uuid, version=None):
+        self.__uuid = object_uuid
+        self.__version = version
+
+    @property
+    def uuid(self):
+        return self.__uuid
+
+    @property
+    def version(self):
+        return self.__version
+
+    def __str__(self):
+        return '%s(%s)' % (type(self).object_type, self.__uuid)
+
+    def unique_label(self):
+        return (type(self).object_type, self.__uuid)
+
+    def add_event(self, operation, phase, duration=None, msg=None):
+        db.add_event(
+            type(self).object_type, self.__uuid, operation, phase, duration, msg)
+
+    @classmethod
+    def _db_create(cls, object_uuid, metadata):
+        metadata['uuid'] = object_uuid
+        etcd.create(cls.object_type, None, object_uuid, metadata)
+
+    @classmethod
+    def _db_get(cls, object_uuid):
+        # NOTE(mikal): we don't do upgrades inflight. They are assumed to have been
+        # done as part of the upgrade process.
+        o = etcd.get(cls.object_type, None, object_uuid)
+        if not o:
+            LOG.withField(cls.object_type, object_uuid).error(
+                'Object missing')
+            return None
+
+        if o.get('version', 0) != 2:
```

Comment:
The 2 should probably be a constant of the class. That would also support a check at the end of `upgrade` to ensure all objects are at the correct version.

Reply:
Done in 1ec7fa414699eed37c333cf97c6a66e53bbab365.